ios/browser/ui/webui/brave_web_ui_controller_factory.mm

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/ios/browser/ui/webui/brave_web_ui_controller_factory.h"
+
+#include <memory>
+
+#include "base/feature_list.h"
+#include "base/memory/ptr_util.h"
+#include "base/no_destructor.h"
+#include "brave/components/constants/pref_names.h"
+#include "brave/components/constants/url_constants.h"
+#include "brave/components/constants/webui_url_constants.h"
+#include "build/build_config.h"
+#include "components/prefs/pref_service.h"
+#include "ios/components/webui/web_ui_url_constants.h"
+#include "url/gurl.h"
+
+#include "ios/chrome/browser/shared/model/browser_state/chrome_browser_state.h"
+#include "ios/chrome/browser/shared/model/url/chrome_url_constants.h"
+
+using web::WebUIIOS;
+using web::WebUIIOSController;
+
+namespace brave {
+
+// A function for creating a new WebUIIOS.
+using WebUIIOSFactoryFunction =
+    std::unique_ptr<WebUIIOSController> (*)(WebUIIOS* web_ui, const GURL& url);
+
+// Template for defining WebUIIOSFactoryFunction.
+template <class T>
+std::unique_ptr<WebUIIOSController> NewWebUIIOS(WebUIIOS* web_ui,
+                                                const GURL& url) {
+  return std::make_unique<T>(web_ui, url);
+}
+
+// Returns a function that can be used to create the right type of WebUIIOS for
+// a tab, based on its URL. Returns nullptr if the URL doesn't have WebUIIOS
+// associated with it.
+WebUIIOSFactoryFunction GetWebUIIOSFactoryFunction(const GURL& url) {
+  // This will get called a lot to check all URLs, so do a quick check of other
+  // schemes to filter out most URLs.
+  if (!url.SchemeIs(kBraveUIScheme) && !url.SchemeIs(kChromeUIScheme)) {
+    return nullptr;
+  }
+
+  // TODO: Handle RewardsInternalUI, AdsInternalUI, AdblockInternalUI URLs here
+  // ChromeBrowserState* browser_state =
+  // ChromeBrowserState::FromWebUIIOS(web_ui);
+
+  const std::string url_host = url.host();
+  /*if (url_host == kAdblockInternalsHost) {
+    return &NewWebUIIOS<BraveAdblockInternalsUI>;
+  } else if (url_host == kSkusInternalsHost) {
+    return &NewWebUIIOS<SkusInternalsUI>;
+  } else if (url_host == kRewardsInternalsHost &&
+             brave_rewards::IsSupportedForProfile(browser_state)) {
+    return &NewWebUIIOS<BraveRewardsInternalsUI>;
+  }*/
```

Comment:
a lot of leftovers here and there

Reply:
I need to keep these ones here. This is because another PR will come that enables Brave WebUI features here and I don't want to forget that some of them require the browser-state and others don't.